File: graph_components/bfs/src/BfsDemo.js
// JavaScript
// グラフの連結成分: BFS

class GraphData {
    constructor() {
        this._data = {};
    }

    get() {
        return this._data;
    }

    getVertices() {
        return Object.keys(this._data);
    }

    getNeighbors(vertex) {
        return this._data[vertex] || null;
    }

    getVertice(vertex) {
        if (vertex in this._data) {
            return this._data[vertex];
        } else {
            console.error(`ERROR: ${vertex}は範囲外です`);
            return null;
        }
    }

    addVertex(vertex) {
        if (!(vertex in this._data)) {
            this._data[vertex] = [];
        }
        return true;
    }

    addEdge(vertex1, vertex2, weight) {
        if (!(vertex1 in this._data)) {
            this.addVertex(vertex1);
        }
        if (!(vertex2 in this._data)) {
            this.addVertex(vertex2);
        }

        const addOrUpdateEdge = (v1, v2) => {
            const existingEdgeIndex = this._data[v1].findIndex(([neighbor]) => neighbor === v2);
            if (existingEdgeIndex !== -1) {
                this._data[v1][existingEdgeIndex] = [v2, weight];
            } else {
                this._data[v1].push([v2, weight]);
            }
        };

        addOrUpdateEdge(vertex1, vertex2);
        addOrUpdateEdge(vertex2, vertex1);

        return true;
    }

    clear() {
        this._data = {};
        return true;
    }

    getConnectedComponents() {
        const visited = new Set();
        const allComponents = [];

        const vertices = this.getVertices();

        for (const vertex of vertices) {
            if (!visited.has(vertex)) {
                const currentComponent = [];
                const queue = [vertex];
                visited.add(vertex);
                currentComponent.push(vertex);

                while (queue.length > 0) {
                    const u = queue.shift();

                    const neighborsWithWeight = this.getNeighbors(u);

                    if (neighborsWithWeight) {
                        for (const [neighbor] of neighborsWithWeight) {
                            if (!visited.has(neighbor)) {
                                visited.add(neighbor);
                                queue.push(neighbor);
                                currentComponent.push(neighbor);
                            }
                        }
                    }
                }

                allComponents.push(currentComponent);
            }
        }

        return allComponents;
    }
}

function main() {
    console.log("Bfs TEST -----> start");

    console.log("\nnew");
    const graphData = new GraphData();
    console.log(`  現在のデータ: ${JSON.stringify(graphData.get())}`);

    console.log("\nadd_edge");
    graphData.clear();
    const inputList1 = [['A', 'B', 4], ['B', 'C', 3], ['B', 'D', 2], ['D', 'A', 1], ['A', 'C', 2], ['B', 'D', 2]];
    for (const input of inputList1) {
        console.log(`  入力値: ${input}`);
        const output = graphData.addEdge(...input);
        console.log(`  出力値: ${output}`);
    }
    console.log(`  現在のデータ: ${JSON.stringify(graphData.get())}`);
    console.log("\nget_connected_components");
    const output1 = graphData.getConnectedComponents();
    console.log(`  連結成分: ${JSON.stringify(output1)}`);

    console.log("\nadd_edge");
    graphData.clear();
    const inputList2 = [['A', 'B', 4], ['C', 'D', 4], ['E', 'F', 1], ['F', 'G', 1]];
    for (const input of inputList2) {
        console.log(`  入力値: ${input}`);
        const output = graphData.addEdge(...input);
        console.log(`  出力値: ${output}`);
    }
    console.log(`  現在のデータ: ${JSON.stringify(graphData.get())}`);
    console.log("\nget_connected_components");
    const output2 = graphData.getConnectedComponents();
    console.log(`  連結成分: ${JSON.stringify(output2)}`);

    console.log("\nadd_edge");
    graphData.clear();        
    const inputList3 = [['A', 'B', 4], ['B', 'C', 3], ['D', 'E', 5]];
    for (const input of inputList3) {
        console.log(`  入力値: ${input}`);
        const output = graphData.addEdge(...input);
        console.log(`  出力値: ${output}`);
    }
    console.log(`  現在のデータ: ${JSON.stringify(graphData.get())}`);
    console.log("\nget_connected_components");
    const output3 = graphData.getConnectedComponents();
    console.log(`  連結成分: ${JSON.stringify(output3)}`);

    console.log("\nadd_edge");
    graphData.clear();
    const inputList4 = [];
    for (const input of inputList4) {
        console.log(`  入力値: ${input}`);
        const output = graphData.addEdge(...input);
        console.log(`  出力値: ${output}`);
    }
    console.log(`  現在のデータ: ${JSON.stringify(graphData.get())}`);
    console.log("\nget_connected_components");
    const output4 = graphData.getConnectedComponents();
    console.log(`  連結成分: ${JSON.stringify(output4)}`);

    console.log("Bfs TEST <----- end");
}

// Uncomment the line below to run the main function
// main();

File: graph_components/bfs/src/BfsDemo.kt
// Kotlin
// グラフの連結成分: BFS

import java.util.*

class GraphData {
    private val _data: MutableMap<String, MutableList<Pair<String, Int>>> = mutableMapOf()

    fun get(): Map<String, List<Pair<String, Int>>> = _data

    fun getVertices(): List<String> = _data.keys.toList()
        
    fun getNeighbors(vertex: String): List<Pair<String, Int>>? {
        return _data[vertex]
    }

    fun getVertice(vertex: String): List<Pair<String, Int>>? {
        return if (vertex in _data) {
            _data[vertex]
        } else {
            println("ERROR: $vertex は範囲外です")
            null
        }
    }

    fun addVertex(vertex: String): Boolean {
        if (vertex !in _data) {
            _data[vertex] = mutableListOf()
            return true
        }
        return true
    }
    
    fun addEdge(vertex1: String, vertex2: String, weight: Int): Boolean {
        if (vertex1 !in _data) {
            addVertex(vertex1)
        }
        if (vertex2 !in _data) {
            addVertex(vertex2)
        }
        
        val v1Edges = _data[vertex1]!!
        val v2Edges = _data[vertex2]!!
        
        val existsV1V2 = v1Edges.indexOfFirst { it.first == vertex2 }.let { index ->
            if (index != -1) {
                v1Edges[index] = Pair(vertex2, weight)
                true
            } else {
                v1Edges.add(Pair(vertex2, weight))
                false
            }
        }
        
        val existsV2V1 = v2Edges.indexOfFirst { it.first == vertex1 }.let { index ->
            if (index != -1) {
                v2Edges[index] = Pair(vertex1, weight)
                true
            } else {
                v2Edges.add(Pair(vertex1, weight))
                false
            }
        }
        
        return true
    }
        
    fun clear(): Boolean {
        _data.clear()
        return true
    }
    
    fun getConnectedComponents(): List<List<String>> {
        val visited = mutableSetOf<String>()
        val allComponents = mutableListOf<List<String>>()

        val vertices = getVertices()

        for (vertex in vertices) {
            if (vertex !in visited) {
                val currentComponent = mutableListOf<String>()
                val queue: Queue<String> = LinkedList()
                queue.add(vertex)
                visited.add(vertex)
                currentComponent.add(vertex)

                while (queue.isNotEmpty()) {
                    val u = queue.poll()

                    val neighborsWithWeight = getNeighbors(u)

                    neighborsWithWeight?.forEach { (neighbor, _) ->
                        if (neighbor !in visited) {
                            visited.add(neighbor)
                            queue.add(neighbor)
                            currentComponent.add(neighbor)
                        }
                    }
                }

                allComponents.add(currentComponent)
            }
        }

        return allComponents
    }
}

fun main() {
    println("Bfs TEST -----> start")

    println("\nnew")
    val graphData = GraphData()
    println("  現在のデータ: ${graphData.get()}")

    println("\nadd_edge")
    graphData.clear()
    val inputList1 = listOf(
        Triple("A", "B", 4), 
        Triple("B", "C", 3), 
        Triple("B", "D", 2), 
        Triple("D", "A", 1), 
        Triple("A", "C", 2), 
        Triple("B", "D", 2)
    )
    for (input in inputList1) {
        println("  入力値: $input")
        val output = graphData.addEdge(input.first, input.second, input.third)
        println("  出力値: $output")
    }
    println("  現在のデータ: ${graphData.get()}")
    println("\nget_connected_components")
    val output1 = graphData.getConnectedComponents()
    println("  連結成分: $output1")

    println("\nadd_edge")
    graphData.clear()
    val inputList2 = listOf(
        Triple("A", "B", 4), 
        Triple("C", "D", 4), 
        Triple("E", "F", 1), 
        Triple("F", "G", 1)
    )
    for (input in inputList2) {
        println("  入力値: $input")
        val output = graphData.addEdge(input.first, input.second, input.third)
        println("  出力値: $output")
    }
    println("  現在のデータ: ${graphData.get()}")
    println("\nget_connected_components")
    val output2 = graphData.getConnectedComponents()
    println("  連結成分: $output2")

    println("\nadd_edge")
    graphData.clear()
    val inputList3 = listOf(
        Triple("A", "B", 4), 
        Triple("B", "C", 3), 
        Triple("D", "E", 5)
    )
    for (input in inputList3) {
        println("  入力値: $input")
        val output = graphData.addEdge(input.first, input.second, input.third)
        println("  出力値: $output")
    }
    println("  現在のデータ: ${graphData.get()}")
    println("\nget_connected_components")
    val output3 = graphData.getConnectedComponents()
    println("  連結成分: $output3")

    println("\nadd_edge")
    graphData.clear()
    val inputList4 = listOf<Triple<String, String, Int>>()
    for (input in inputList4) {
        println("  入力値: $input")
        val output = graphData.addEdge(input.first, input.second, input.third)
        println("  出力値: $output")
    }
    println("  現在のデータ: ${graphData.get()}")
    println("\nget_connected_components")
    val output4 = graphData.getConnectedComponents()
    println("  連結成分: $output4")

    println("Bfs TEST <----- end")
}

// プログラムのエントリーポイント
fun main(args: Array<String>) {
    main()
}
